pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,270 @@
+		return nil, fmt.Errorf("could not parse jwt payload: %w", err)
+	}
+
+	return &JwtX509Token{
+		chain:  chain,
+		token:  parsedJwt,
+		raw:    rawAuthToken,
+		sigAlg: headers.Algorithm(),
+	}, nil
+}
+
+func (validator JwtX509Validator) parseCertsFromHeader(certsFromHeader []string) ([]*x509.Certificate, error) {
+	// Parse all the certificates from the header into a chain
+	chain := []*x509.Certificate{}
+
+	for _, certStr := range certsFromHeader {
+		rawCert, err := base64.StdEncoding.Strict().DecodeString(certStr)
+		if err != nil {
+			return nil, fmt.Errorf("could not base64 decode cert: %w", err)
+		}
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse x509 certificate: %w", err)
+		}
+		chain = append(chain, cert)
+	}
+	return chain, nil
+}
+
+func (validator JwtX509Validator) Verify(x509Token *JwtX509Token) error {
+	var sigAlgAllowed bool
+	for _, allowedAlg := range validator.allowedSigningAlgs {
+		if allowedAlg == x509Token.sigAlg {
+			sigAlgAllowed = true
+			break
+		}
+	}
+	if !sigAlgAllowed {
+		return fmt.Errorf("signature algorithm %s is not allowed", x509Token.sigAlg)
+	}
+
+	leafCert, verifiedChains, err := validator.verifyCertChain(x509Token)
+	if err != nil {
+		return err
+	}
+
+	for _, verifiedChain := range verifiedChains {
+		err := validator.checkCertRevocation(verifiedChain)
+		if err != nil {
+			return err
+		}
+	}
+
+	// parse the jwt and verify the jwt signature
+	token, err := jwt.ParseString(x509Token.raw, jwt.WithVerify(x509Token.sigAlg, leafCert.PublicKey))
+	if err != nil {
+		return err
+	}
+
+	// verify the jwt contents such as expiration
+	if err := jwt.Verify(token); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (validator JwtX509Validator) verifyCertChain(x509Token *JwtX509Token) (*x509.Certificate, [][]*x509.Certificate, error) {
+	rootsPool := x509.NewCertPool()
+	for _, cert := range validator.roots {
+		rootsPool.AddCert(cert)
+	}
+	intermediatesPool := x509.NewCertPool()
+	for _, cert := range validator.intermediates {
+		intermediatesPool.AddCert(cert)
+	}
+	verifyOpts := x509.VerifyOptions{
+		Intermediates: intermediatesPool,
+		Roots:         rootsPool,
+		CurrentTime:   time.Now(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	if len(x509Token.chain) == 0 {
+		return nil, nil, fmt.Errorf("token does not have a certificate")
+	}
+
+	for _, cert := range x509Token.chain[1:len(x509Token.chain)] {
+		verifyOpts.Intermediates.AddCert(cert)
+	}
+
+	leafCert := x509Token.chain[0]
+
+	verifiedChains, err := leafCert.Verify(verifyOpts)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to verify certificate chain: %w", err)
+	}
+
+	rootCert := verifiedChains[0][len(verifiedChains[0])-1]
+	if !bytes.Equal(rootCert.RawIssuer, rootCert.RawSubject) || rootCert.CheckSignatureFrom(rootCert) != nil {
+		return nil, nil, fmt.Errorf("certificate is not a root CA")
```

Comment:
no formatting directive here. Maybe also include the subject/issuer in the error message?